dhcp-protocol/src/main/java/org/anarres/dhcp/v6/io/Dhcp6MessageEncoder.java

```diff
@@ -28,23 +29,53 @@ public static Dhcp6MessageEncoder getInstance() {
         return Inner.INSTANCE;
     }
 
+    /**
+     * https://tools.ietf.org/html/rfc3315#section-6
+     * https://tools.ietf.org/html/rfc3315#section-7
+     */
     public void encode(ByteBuffer byteBuffer, Dhcp6Message message) {
         byteBuffer.put(message.getMessageType().getCode());
-        int transactionId = message.getTransactionId();
-        byteBuffer.put((byte) ((transactionId >> 16) & 0xFF));
-        byteBuffer.put((byte) ((transactionId >> 8) & 0xFF));
-        byteBuffer.put((byte) (transactionId & 0xFF));
+        if(Dhcp6MessageDecoder.isRelayedMessage(message.getMessageType())) {
+            final Dhcp6RelayMessage relayedMessage = (Dhcp6RelayMessage) message;
+            byteBuffer.put(relayedMessage.getHopCount());
+            byteBuffer.put(relayedMessage.getLinkAddress().getAddress());
+            byteBuffer.put(relayedMessage.getPeerAddress().getAddress());
+        } else {
+            int transactionId = message.getTransactionId();
+            byteBuffer.put((byte) ((transactionId >> 16) & 0xFF));
+            byteBuffer.put((byte) ((transactionId >> 8) & 0xFF));
+            byteBuffer.put((byte) (transactionId & 0xFF));
+        }
         encode(byteBuffer, message.getOptions());
     }
 
     public void encode(@Nonnull ByteBuffer byteBuffer, @Nonnull Dhcp6Options options) {
         for (Dhcp6Option option : options) {
             // Option continuation per RFC3396
-            short tag = option.getTag();
-            byte[] data = option.getData();
-            byteBuffer.putShort(tag);
-            byteBuffer.putShort(Shorts.checkedCast(data.length));
-            byteBuffer.put(data);
+            encode(byteBuffer, option);
+        }
+    }
+
+    public ByteBuffer encode(@Nonnull Dhcp6Options options) {
+        ByteBuffer allOptions = ByteBuffer.allocate(0);
```

Comment:
I've spent far too much time in the HFT community... it would be lovely to preallocate this, even at the cost of iterating options twice, so we don't have reallocation in the loop.